### ISSUE TITLE
Backpressure-Handling of WorkQueueProcessor leads to unnecessarily dropped messages with bursty producers

### ISSUE THREAD

**Issue by mlex** [reactor/reactor-core#104]
## TL;DR
Because the following explanation are a bit long, I'll try to sum it up in a few bullet points:
* I use a WorkQueueProcessor with RingBuffer-capacity 128.
* There are never more than 15 slots occupied (WorkQueueProcessor.getPending() is always below 15).
* The consumer faster than the producer.
* But still: **Backpressure drops messages!**

## Details
In the following situation, the producer publishes messages with a rate of 200 messages per second. It does so by always publishing bursts of 5 messages and then sleeping for 25ms.

The final consumer consumes the messages with a rate of 250 message per second.

To buffer the bursty producer, I wanted to use a workqueueprocessor with a capacity of 128. In the test it is initially filled with 10 messages (in a real application, this could be e.g. a remainder of a previous burst).

Using Backpressure strategy DROP, I would expect that no messages at all are dropped, because the consumer is much faster than the producer and the ringbuffer should take care of possible lags and bursts.

But actually after the first ~880 messages, there appear regular DROPs! To illustrate the problem, I created a small unit-test (see https://gist.github.com/mlex/a7349216f32c4f94acc0c9d7c10d6321 for the full test):

```java
  public static final int PRODUCER_LATENCY = 5;
  public static final int CONSUMER_LATENCY = 4;

  public static final int RINGBUFFER_SIZE = 128;

  public static final int INITAL_MESSAGES_COUNT   = 10;
  public static final int PRODUCED_MESSAGES_COUNT = 1024;
  public static final int BURST_SIZE              = 5;

  @Test
  public void test() throws Exception {
    processor = WorkQueueProcessor.create("test-processor", RINGBUFFER_SIZE);

    Flux
      .create((emitter) -> burstyProducer(emitter, PRODUCED_MESSAGES_COUNT, BURST_SIZE))
      .onBackpressureDrop(this::incrementDroppedMessagesCounter)
      .subscribe(processor);

    Flux.from(processor)
      .map(this::complicatedCalculation)
      .subscribe(this::logConsumedValue);

    waitForProducerFinish();

    System.out.println("\n\nMax ringbuffer pending: " + maxRingBufferPending.get());

    assertEquals(0, getDroppedMessagesCount(), "Expect zero dropped messages");
  }
```

## Possible explanation

The reason for the observed behaviour lies in the way `Subscription.request()` is called in `reactor.queue.RequestTask` (which is used by `WorkQueueProcessor`). The `RequestTask` initially requests `bufferSize-1` messages (in our case 127). It then waits for the ringbuffer to consume `bufferSize*3/4` (in our case 96) messages and then requests another `bufferSize*3/4` messages.

As far as I can see, the intention behind this is the following: In the case of a fast producer and slow consumer, this would lead to a ringbuffer, that is always filled with 31 to 127 elements. In case of a slow producer and fast consumer, this would lead to a number of pending requests being constantly between 31 and 127 elements.

Unfortunately the condition `waitStrategy.waitFor(cursor + limit, readCount, spinObserver);` (see [RingBuffer.java:868](https://github.com/reactor/reactor-core/blob/0d459836c9b3621846bedc48a3b839e465167726/src/main/java/reactor/core/queue/RingBuffer.java#L868)) that should wait for the next `limit`(=96) elements to be processed, is not triggered after exactly 96 elements. It is triggered after *at least* 96 elements, because the `signalAllWhenBlocking()` in `WorkQueueProcessor` is only called after `nextSequence > cachedAvailableSequence` is reached (see [WorkQueueProcessor.java:818](https://github.com/reactor/reactor-core/blob/0d459836c9b3621846bedc48a3b839e465167726/src/main/java/reactor/core/publisher/WorkQueueProcessor.java#L818)).

In the case of bursty producer, the `cachedAvailableSequence` does not increase constantly, but makes jumps (the size of the jumps being the burstSize). In the above example, this causes `waitFor` to return after 100 elements have been processed (not after 96 as expected)! So we actually "loose" 4 slots ("loose", because they will never be filled again - even if the producer starts producing at a much faster rate). 

After 800 elements, we have already lost 8*4 = 32 slots. This means, that the next `Subscription.request()` is only triggered when the number of pending requests equals zero - in which case we start dropping messages.

**Comment by smaldini** [comment]
Excellent report @mlex this is giving quite a lead to fix that asap. 
Somehow the new `Flux.parallel` could give you a similar "starve the cpu with N concurrent sub" behavior. However it's a fixed concurrency size in that case vs `Processor` creating a `Thread` per `Subscriber`. Do you have a need for dynamic subscribers ?

**Comment by smaldini** [comment]
Forgot to ask which version is that ? I don't think it relates to WorkProcessor much but `Flux#parallel` is new to 2.5.0.M4, and the `Flux.create` signature has changed.

**Comment by mlex** [comment]
I wrote the Unit-Test using the latest reactor-core master (commit 49b899b4).

Thanks for the tip with `Flux#parallel`. I didn't know this exists. We don't need dynamic subscribers, so this would fit. I also like, that in `Flux#parallel' the `prefetch` count is configurable. 

When looking at the code of `Flux#parallel` I noticed, that there is also this logic, that `Subscriber.requests()` is always called after `prefetch * 3/4` elements (e.g.  [ParallelUnorderedRunOn.java#122](https://github.com/reactor/reactor-core/blob/aefbe5303725ca9c6aec172a86d672e16a430a9b/src/main/java/reactor/core/publisher/ParallelUnorderedRunOn.java#L122). Is there a reason for this `3/4`?

**Comment by smaldini** [comment]
Most of the time our buffering operators (flatmap, groupBy, publishOn, parallel...) will accept prefetch and will use that prefetch with 75% threshold for replenishing.
The intent is to avoid the "micro pause" introduced when you wait until all request volume has been received: Calling request will cost usually and is not instant . Hard to determine how much it will cost for sure but in general it overcomes simple async hand off (async request). This a magic number that can't be tweaked right now but we can look at options to tune it, however it might be a global env option at first to avoid explosion of signatures methods in Flux/Mono. 
It seems over time we might encourage only parallel and get rid of WorkQueueProcessor in reactor-core (might be an extension as its niche has some uses still, the dynamic aspect being one).

**Comment by smaldini** [comment]
That would be great if you could confirm the resolution, I added with your credit the gist you linked to verify this on our CI as well.

**Comment by smaldini** [comment]
Actually updated the algo to adapt the request directly https://github.com/reactor/reactor-core/commit/43fb98022a528f5cd6417f58e3a578d77a74d68b#diff-5d65c4f41e3dd9b42ed52c0acdb2c53cR566
